src/nsrt_learning/strips_learning/clustering_learner.py

```diff
@@ -0,0 +1,234 @@
+        # Run hill climbing search, starting from original PNADs.
+        path, _, _ = utils.run_hill_climbing(
+            tuple(pnads), self._check_goal, self._get_sidelining_successors,
+            functools.partial(self._evaluate, pnads))
+        # The last state in the search holds the final PNADs.
+        pnads = list(path[-1])
+        # Because the PNADs have been modified, recompute the datastores.
+        self._recompute_datastores_from_segments(pnads)
+        return pnads
+
+    @abc.abstractmethod
+    def _evaluate(self, initial_pnads: List[PartialNSRTAndDatastore],
+                  s: Tuple[PartialNSRTAndDatastore, ...]) -> float:
+        """Abstract evaluation/score function for search.
+
+        Lower is better.
+        """
+        raise NotImplementedError("Override me!")
+
+    @staticmethod
+    def _check_goal(s: Tuple[PartialNSRTAndDatastore, ...]) -> bool:
+        del s  # unused
+        # There are no goal states for this search; run until exhausted.
+        return False
+
+    @staticmethod
+    def _get_sidelining_successors(
+        s: Tuple[PartialNSRTAndDatastore, ...],
+    ) -> Iterator[Tuple[None, Tuple[PartialNSRTAndDatastore, ...], float]]:
+        # For each PNAD/operator...
+        for i in range(len(s)):
+            pnad = s[i]
+            _, option_vars = pnad.option_spec
+            # ...consider changing each of its add effects to a side predicate.
+            for effect in pnad.op.add_effects:
+                if len(pnad.op.add_effects) > 1:
+                    # We don't want sidelining to result in a no-op.
+                    new_pnad = PartialNSRTAndDatastore(
+                        pnad.op.effect_to_side_predicate(
+                            effect, option_vars, "add"), pnad.datastore,
+                        pnad.option_spec)
+                    sprime = list(s)
+                    sprime[i] = new_pnad
+                    yield (None, tuple(sprime), 1.0)
+
+            # ...consider removing it.
+            sprime = list(s)
+            del sprime[i]
+            yield (None, tuple(sprime), 1.0)
+
+
+class ClusterAndIntersectSidelinePredictionErrorSTRIPSLearner(
+        ClusterAndIntersectSidelineSTRIPSLearner):
+    """A STRIPS learner that uses hill climbing with a prediction error score
+    function for side predicate learning."""
+
+    def _evaluate(self, initial_pnads: List[PartialNSRTAndDatastore],
+                  s: Tuple[PartialNSRTAndDatastore, ...]) -> float:
+        segments = [seg for traj in self._segmented_trajs for seg in traj]
+        strips_ops = [pnad.op for pnad in s]
+        option_specs = [pnad.option_spec for pnad in s]
+        num_true_positives, num_false_positives, _, _ = \
+            utils.count_positives_for_ops(strips_ops, option_specs, segments)
+        # Note: lower is better! We want more true positives and fewer
+        # false positives.
+        return num_false_positives + 10 * (-num_true_positives)
```

Comment:
we could make this a hyperparameter (we were previously using `grammar_search_{true,false}_pos_weight` but i think this is better) but do we really need more?

Reply:
i prefer it as a hyperparameter, so that we can more easily look at settings.py in the relevant section and remember all the hyperparameters that we should be reporting in a paper